### PR TITLE
feat(behavior_path_planner): add always executable module option

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

add always executable module option and use it for fixed_goal_planner

## Tests performed

psim

https://evaluation.tier4.jp/evaluation/reports/tables/new?catalog_id=cb497016-2e2f-4ce0-9ac8-68d77a27a55d&filter=&job_id=fcd5b992-96de-5b41-a439-9741a5ac9848&project_id=prd_jt&table_config=date%2Chide&target_ids=8b5e813f-8274-511e-afc1-addd20898a79

when avoidance is not 'simultaneously executable'.

image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 1]
```diff
@@ -206,7 +206,12 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
       std::find_if(approved_module_ptrs_.begin(), approved_module_ptrs_.end(), find_block_module);
 
     if (itr != approved_module_ptrs_.end()) {
-      return {};
+      const bool has_always_executable_approved_module = std::any_of(
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:209]: It should use `std::all_of`...?

**Comment by kosuke55** [comment]
@satoshi-ota 
thanks! updated condtions and comments in https://github.com/autowarefoundation/autoware.universe/pull/4785/commits/b1fbbadcbcf3cd6ec1ced17ce15b83ac15e487e2

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 2]
```diff
@@ -218,14 +202,47 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
...
-      if (!manager_ptr->isSimultaneousExecutableAsApprovedModule()) {
+    const bool has_non_always_executable_module = std::any_of(
+      approved_module_ptrs_.begin(), approved_module_ptrs_.end(),
+      [this](const auto & m) { return !getManager(m)->isAlwaysExecutableModule(); });
+    if (has_non_always_executable_module) {
+      // Condition 1: do not add modules that are neither always nor simultaneous executable
+      // if there exists at least one approved module that is simultaneous but not always
+      // executable. (only modules that are either always executable or simultaneous executable can
+      // be added)
+      const auto find_non_always_simultaneous_module = [this](const auto & m) {
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:215]:
```suggestion
      const auto find_block_module = [this](const auto & m) {
```

[imo] `find_non_always_simultaneous_module` is not understandable. I think it is better to use `find_block_module`.

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 3]
```diff
@@ -218,14 +202,47 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
...
+        return !getManager(m)->isAlwaysExecutableModule() &&
+               getManager(m)->isSimultaneousExecutableAsApprovedModule();
+      };
+      const auto itr_non_always_simultaneous = std::find_if(
+        approved_module_ptrs_.begin(), approved_module_ptrs_.end(),
+        find_non_always_simultaneous_module);
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:221]:
```suggestion
        find_block_module);
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 4]
```diff
@@ -218,14 +202,47 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
...
+    if (has_non_always_executable_module) {
+      // Condition 1: do not add modules that are neither always nor simultaneous executable
+      // if there exists at least one approved module that is simultaneous but not always
+      // executable. (only modules that are either always executable or simultaneous executable can
+      // be added)
+      const auto find_non_always_simultaneous_module = [this](const auto & m) {
+        return !getManager(m)->isAlwaysExecutableModule() &&
+               getManager(m)->isSimultaneousExecutableAsApprovedModule();
+      };
+      const auto itr_non_always_simultaneous = std::find_if(
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:219]:
```suggestion
      const auto itr_block = std::find_if(
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 5]
```diff
@@ -218,14 +202,47 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
...
+        approved_module_ptrs_.begin(), approved_module_ptrs_.end(),
+        find_non_always_simultaneous_module);
+      if (
+        itr_non_always_simultaneous != approved_module_ptrs_.end() &&
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:223]:
```suggestion
        itr_block != approved_module_ptrs_.end() &&
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 6]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
+      executable_modules.push_back(module_ptr);
+      continue;
+    }
+
+    // Condition 2: Only modules that are always executable can be added
+    // if there exists at least one executable module that is neither always nor simultaneous
+    // executable.
+    const auto find_block_module = [this](const auto & m) {
+      return !getManager(m)->isAlwaysExecutableModule() &&
+             !getManager(m)->isSimultaneousExecutableAsApprovedModule();
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:388]:
```suggestion
             !getManager(m)->isSimultaneousExecutableAsCandidateModule();
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 7]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
+      continue;
+    }
 
-    if (itr == executable_modules.end()) {
+    // Condition 1: Only modules that are either always executable or simultaneous executable can be
+    // added if there exists at least one executable module that is simultaneous but not always
+    // executable.
+    const auto find_non_always_simultaneous_module = [this](const auto & m) {
+      return !getManager(m)->isAlwaysExecutableModule() &&
+             getManager(m)->isSimultaneousExecutableAsApprovedModule();
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:403]:
```suggestion
             getManager(m)->isSimultaneousExecutableAsCandidateModule();
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 8]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
+    };
+    const auto itr_non_always_simultaneous = std::find_if(
+      executable_modules.begin(), executable_modules.end(), find_non_always_simultaneous_module);
+    if (
+      itr_non_always_simultaneous != executable_modules.end() &&
+      (getManager(module_ptr)->isAlwaysExecutableModule() ||
+       getManager(module_ptr)->isSimultaneousExecutableAsApprovedModule())) {
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:410]:
```suggestion
       getManager(module_ptr)->isSimultaneousExecutableAsCandidateModule())) {
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 9]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
       executable_modules.push_back(module_ptr);
+      continue;
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:412]: necessary?

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/planner_manager.cpp:215]: ok!
used `find_block_module` also in Condition 2, so add unnamed namespace?

```
    // Condition 2: Only modules that are always executable can be added
    // if there exists at least one executable module that is neither always nor simultaneous
    // executable.
    const auto find_block_module = [this](const auto & m) {
      return !getManager(m)->isAlwaysExecutableModule() &&
             !getManager(m)->isSimultaneousExecutableAsCandidateModule();
    };
```

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 10]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
   for (const auto & module_ptr : sorted_request_modules) {
-    if (!getManager(module_ptr)->isSimultaneousExecutableAsCandidateModule()) {
-      if (executable_modules.empty()) {
-        executable_modules.push_back(module_ptr);
-        break;
-      }
+    // if this module is always executable, add it to the list immediately.
+    if (getManager(module_ptr)->isAlwaysExecutableModule()) {
+      executable_modules.push_back(module_ptr);
+      continue;
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/planner_manager.cpp]: break?

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 11]
```diff
@@ -347,20 +364,52 @@ std::pair<SceneModulePtr, BehaviorModuleOutput> PlannerManager::runRequestModule
...
+    const auto find_block_module = [this](const auto & m) {
+      return !getManager(m)->isAlwaysExecutableModule() &&
+             !getManager(m)->isSimultaneousExecutableAsCandidateModule();
+    };
+    const auto itr_block =
+      std::find_if(executable_modules.begin(), executable_modules.end(), find_block_module);
+    if (
+      itr_block != executable_modules.end() && getManager(module_ptr)->isAlwaysExecutableModule()) {
+      executable_modules.push_back(module_ptr);
+      continue;
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/planner_manager.cpp]: break

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 9 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/planner_manager.cpp]: no, fixed in [0071bbd](https://github.com/autowarefoundation/autoware.universe/pull/4785/commits/0071bbdd8aa4d510a67f6d9bc32e23716f7aa416)

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 12]
```diff
@@ -218,14 +202,62 @@ std::vector<SceneModulePtr> PlannerManager::getRequestModules(
...
+        }
+      }
+      if (skip_module) {
         continue;
       }
     }
+    // else{
+    //   Condition 4: if none of the above conditions are met, any module can be added.
+    //   (when the approved modules are either empty or consist only of always executable modules.)
+    // }
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:260]: @kosuke55 You didn't remove this comment out intentionally?

**Review by satoshi-ota — APPROVED** [review]

File: planning/behavior_path_planner/src/planner_manager.cpp (hunk 12 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/planner_manager.cpp:260]: yes!

**Pull request merged**